File: fastapi-async/src/services/auth_service.py
from databases import Database

from resources.security import validate, encode_token_jwt
from services.user_service import UserService


class AuthService:
    def __init__(self, database: Database, user_service: UserService):
        self._database: Database = database
        self._user_service: UserService = user_service

    async def authenticate(self, email, password) -> str:
        user = await self._user_service.get_by_email(email)
        if user is None or not validate(password, user["password"]):
            raise ValueError("Invalid username or password")

        return self.generate_token(user)

    def generate_token(self, user: dict):
        token: str = encode_token_jwt(user["email"])

        return token

File: aspnet5/src/Services/AuthService.cs
using System;
using System.Collections.Generic;
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;
using System.Text;
using System.Threading.Tasks;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using Microsoft.IdentityModel.Tokens;
using src.Exceptions;
using src.Models;
using src.Resources;

namespace src.Services
{
    public class AuthService
    {
        private readonly AuthenticationConfiguration _configuration;
        private readonly ILogger<AuthService> _logger;
        private readonly UserService _service;

        public AuthService(IOptions<AuthenticationConfiguration> options, ILogger<AuthService> logger, UserService service)
        {
            _configuration = options.Value;
            _logger = logger;
            _service = service;
        }

        public async Task<string> AuthenticateAsync(string email, string password)
        {
            var user = await _service.GetByEmailAsync(email);
            if (user is null || user.Password != _service.EncryptPassword(password))
                throw new UnauthorizeException(email);

            return GenerateToken(user);
        }

        private string GenerateToken(User user)
        {
            var tokenHandler = new JwtSecurityTokenHandler();
            byte[] key = Encoding.ASCII.GetBytes(_configuration.SecretKey);

            var securityToken = new SecurityTokenDescriptor
            {
                Subject = new ClaimsIdentity(new Claim[]
                {
                    new Claim("id", user.Id.ToString()),
                    new Claim("email", user.Email),
                }),
                Expires = DateTime.UtcNow.AddMinutes(_configuration.MinutesToExpire),
                SigningCredentials = new SigningCredentials(new SymmetricSecurityKey(key), SecurityAlgorithms.HmacSha256Signature)
            };

            var tokenDescriptor = tokenHandler.CreateToken(securityToken);
            
            return tokenHandler.WriteToken(tokenDescriptor);
        }
    }
}